internal/client/db/db.go

```diff
@@ -0,0 +1,585 @@
+		NumberLeaves: bdb.leaves.Count(),
+		BlockGap:     bdb.meta.BlockGap,
+	}
+	return info
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Status(hash H) (blockchain.BlockStatus, error) {
+	header, err := bdb.Header(hash)
+	if err != nil {
+		return 0, err
+	}
+	if header != nil {
+		return blockchain.BlockStatusInChain, nil
+	}
+	return blockchain.BlockStatusUnknown, nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Number(hash H) (*N, error) {
+	meta, err := bdb.HeaderMetadata(hash)
+	if err != nil {
+		return nil, err
+	}
+	return &meta.Number, nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Hash(number N) (*H, error) {
+	header, err := readHeader[H, N, Header](
+		bdb.db,
+		columns.KeyLookup,
+		columns.Header,
+		generic.BlockIDNumber[N]{Inner: number},
+	)
+	if err != nil {
+		return nil, err
+	}
+	if header == nil {
+		return nil, nil
+	}
+	h := (*header).Hash()
+	return &h, nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) BlockHashFromID(id generic.BlockID) (*H, error) {
+	switch id := id.(type) {
+	case generic.BlockIDHash[H]:
+		return &id.Inner, nil
+	case generic.BlockIDNumber[N]:
+		return bdb.Hash(id.Inner)
+	default:
+		panic("unsupported block id type")
+	}
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) BlockNumberFromID(id generic.BlockID) (*N, error) {
+	switch id := id.(type) {
+	case generic.BlockIDHash[H]:
+		return bdb.Number(id.Inner)
+	case generic.BlockIDNumber[N]:
+		return &id.Inner, nil
+	default:
+		panic("unsupported block id type")
+	}
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Body(hash H) ([]runtime.Extrinsic, error) {
+	bdb.pinnedBlocksCacheMtx.RLock()
+	defer bdb.pinnedBlocksCacheMtx.RUnlock()
+	body := bdb.pinnedBlocksCache.Body(hash)
+	if body != nil {
+		return *body, nil
+	}
+
+	return bdb.bodyUncached(hash)
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Justifications(hash H) (runtime.Justifications, error) {
+	bdb.pinnedBlocksCacheMtx.RLock()
+	defer bdb.pinnedBlocksCacheMtx.RUnlock()
+	justifications := bdb.pinnedBlocksCache.Justifications(hash)
+	if justifications != nil {
+		return justifications, nil
+	}
+
+	return bdb.justificationsUncached(hash)
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) LastFinalized() (H, error) {
+	bdb.metaMtx.RLock()
+	defer bdb.metaMtx.RUnlock()
+	return bdb.meta.FinalizedHash, nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Leaves() ([]H, error) {
+	bdb.leavesMtx.RLock()
+	defer bdb.leavesMtx.RUnlock()
+	return bdb.leaves.Hashes(), nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) DisplacedLeavesAfterFinalizing(blockNumber N) ([]H, error) {
+	bdb.leavesMtx.RLock()
+	defer bdb.leavesMtx.RUnlock()
+	return bdb.leaves.DisplacedByFinalHeight(blockNumber).Leaves(), nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Children(parentHash H) ([]H, error) {
+	return readChildren[H](bdb.db, columns.Meta, metakeys.ChildrenPrefix, parentHash)
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) LongestContaining(baseHash H, importLock *sync.RWMutex) (*H, error) {
```

Comment:
I am thinking that this function should be removed or at the very least modified to remove the import lock being passed in as an argument. I would prefer services handle their locking internally rather than externally, so I think in this case it would be more ideal to have the locking be handled by the import service itself, and then the import service can either retrieve this logic directly after it handles locking internally, or call a version of this function that does not take a mutex as an argument.

Giving that this function is not used at all, that leads me to think we should remove it for now and then figure out a nicer solution for when there is a use case for it.

Reply:
I left this in since we should actually stop any block importing during this process.  See the comments on line 468.  I can revisit this in future work to see if this is the right approach.  I think I can make this a private method with the actual locking mechanism and provide a public `LongestContaining` without the `importLock` in future work when implementing the `db.Backend`.

Reply:
I don't love passing the lock, but I'm okay with not addressing in this PR. I think then it should be added to an issue to keep track of tests to do as part of `db.Backend` or to do after the completion of `db.Backend` to make sure we address everything before its integrated/merged into dev

Reply:
Yup keep an eye out for this change in the PR for #3900.  I agree the public facing method of `db.Backend` will not have the lock.  Keep in mind `blockchainDB` is never exported or exposed to users of the `db` package.